src/main/java/com/xebia/visualreview/PixelComparator.java

```diff
@@ -21,77 +21,149 @@
-
-  public static DiffReport processImage(File beforeFile, File afterFile) {
-    try {
-      PixelGrabber beforeGrab = grabImage(beforeFile);
-      PixelGrabber afterGrab = grabImage(afterFile);
-
-      int[] beforeData = null;
-      int[] afterData = null;
-      int beforeWidth = 0;
-      int afterWidth = 0;
-      int y1 = 0;
-      int y2 = 0;
-
-      if (beforeGrab.grabPixels()) {
-        beforeWidth = beforeGrab.getWidth();
-        y1 = beforeGrab.getHeight();
-        beforeData = (int[]) beforeGrab.getPixels();
-      }
-
-      if (afterGrab.grabPixels()) {
-        afterWidth = afterGrab.getWidth();
-        y2 = afterGrab.getHeight();
-        afterData = (int[]) afterGrab.getPixels();
-      }
-
-      int minX = Math.min(beforeWidth, afterWidth);
-      int minY = Math.min(y1, y2);
-      int diffWidth = Math.max(beforeWidth, afterWidth);
-      int diffHeight = Math.max(y1, y2);
-      int[] diffData = new int[diffWidth * diffHeight];
-      int differentPixels = 0;
-
-      for (int y = 0; y < diffHeight; y++) {
-        for (int x = 0; x < diffWidth; x++) {
-          if (x >= minX || y >= minY || beforeData[y * beforeWidth + x] != afterData[y * afterWidth + x]) {
-              diffData[y * diffWidth + x] = DIFFCOLOUR;
-              differentPixels++;
-          }
+    private static int DIFFCOLOUR = (128<<24) & 0xff000000|       //a
+            (128<<16) &   0xff0000|       //r
+            (0  <<8 ) &     0xff00|       //g
+            (0  <<0 ) &       0xff;       //b
+
+    private static class Rect{
+        private int x;
+        private int y;
+        private int width;
+        private int height;
+        public boolean isValid;
+        public Rect(int x,int y,int width, int height){
+            this.x = x;
+            this.y = y;
+            this.width = width;
+            this.height = height;
+            isValid = true;
         }
-      }
+        public Rect(java.util.Map info, int maxWidth, int maxHeight){
+            isValid = true;
+            try{
+                int x = Integer.parseInt(info.get(Keyword.find("x")).toString());
+                int y = Integer.parseInt(info.get(Keyword.find("y")).toString());
+                int width = Integer.parseInt(info.get(Keyword.find("width")).toString());
+                int height = Integer.parseInt(info.get(Keyword.find("height")).toString());
 
-      BufferedImage diffImage = new BufferedImage(diffWidth, diffHeight, BufferedImage.TYPE_INT_ARGB);
-      diffImage.setRGB(0, 0, diffWidth, diffHeight, diffData, 0, diffWidth);
+                this.x = Math.min(Math.max(0,x),maxWidth);
+                this.y = Math.max(Math.min(this.y,maxHeight),y);
+                this.width = Math.min(Math.max(0,width),maxWidth-this.x);
+                this.height = Math.min(Math.max(0,height),maxHeight-this.y);
+            }catch (Exception e) {
+                System.out.println("Invalid rect "+info);
```

Comment:
I'd much rather have the server throw an error and abort everything instead of failing silently without applying a mask. 

When using VisualReview in a team or set of teams, checking server logs to see why a test doesn't perform as it should is less than ideal. In such a use case, the server should throw an error as quickly as possible to prevent any confusion of the user.

The current way for `PixelComparator` to report errors is to throw a RuntimeException. Not the most elegant way of doing things (ideally it should be a specific exception type), but doable for now.